google/api/resources.proto

```diff
@@ -46,6 +46,15 @@ message ResourceSet {
   // If omitted, this is the name of the message.
   string name = 1;
 
-  // The distinct resources that make up the set.
-  repeated string resource_types = 2;
+  // Component resources that are part of the set.
+  // Resources declared within a resource set must have `name` set.
+  repeated Resource resources = 2;
+
+  // A reference to existing resources (messages or resource definitions)
```

Comment:
May I suggest a slight re-wording:
```
// References to existing resources (messages of resource definitions)
// that are part of the set.
// These may be specified as fully-qualified (e.g. "google.pubsub.v1.Topic")
// or just the resource/proto name if it is defined within the same package.
// The final set of resources in this resource-set is the union of
// `resources` and `resource_references`.
```

This makes it clear that:
* Multiple references may be included here.
* Both fully-qualified and name-only are fine.
* The final set consists of the union of `resources` and `resource_references`.